### PR TITLE
refactor: all the debug logs extracted to a new separate file

### LINKED ISSUE #533: Extract all debug logs texts to a separate file
#### Reason/Context
We have `debug` flag that enables logging of some additional info for the user. Log texts are "hardcoded" in code and therefore again hardcoded in tests.

#### Description
Move log message texts to a separate file and then import log messages in code and tests. This way, when you change or improve a log message, you do it in one file and do not touch tests at all

### PR DESCRIPTION
**Pull request by vishesh13byte**
**Description**

- ...
- ...
- ...

**Related issue(s)**
 Fixes #533

### CODE REVIEW

**Comment by abhijit-hota** [comment]
Good work! I guess the cases can be made more consistent? `UPPER_SNAKE_CASE` for constants and `camelCase` for functions is a good way as done [here](https://github.com/asyncapi/asyncapi-react/blob/master/library/src/constants.ts).

**Comment by vishesh13byte** [comment]
> Good work! I guess the cases can be made more consistent? `UPPER_SNAKE_CASE` for constants and `camelCase` for functions is a good way as done [here](https://github.com/asyncapi/asyncapi-react/blob/master/library/src/constants.

Okay ,I will try to follow that convention .Thanks for your suggestion . Also ,I have made same errors and just trying to solve them.

**Comment by abhijit-hota** [comment]
@vishesh13byte Cool. 

Please note that I'm not a maintainer/collaborator though😅. I was looking at this issue earlier so I checked the PR and thought the casing should be consistent.

**Comment by vishesh13byte** [comment]
> @vishesh13byte Cool.
> 
> Please note that I'm not a maintainer/collaborator though. I was looking at this issue earlier so I checked the PR and thought the casing should be consistent.

Yup , I know buddy .Your suggestion was good and correct , I am just new to contributing to a big org and realizing that convention should be consistent .

**Comment by aayushmau5** [comment]
Two suggestion:
1. Lint your files before commiting using `npm run lint` because I'm seeing some inconsistent whitespace being added.
2. Add good commit messages. Take a look at [this guide](https://github.com/asyncapi/generator/blob/master/CONTRIBUTING.md)

**Comment by vishesh13byte** [comment]
> Two suggestion:
> 
> 1. Lint your files before commiting using `npm run lint` because I'm seeing some inconsistent whitespace being added.
> 2. Add good commit messages. Take a look at [this guide](https://github.com/asyncapi/generator/blob/master/CONTRIBUTING.md)

I had run that command 'npm run lint' and I corrected all the indentations and line paddings ,there was only one warning (not an error) about module exports . Nonetheless ,thanks for your suggestions and I will try to add better commit messages .

Thank you for your concerns

File: package.json [hunk 1]
```diff
@@ -4,8 +4,8 @@
   "description": "The AsyncAPI generator. It can generate documentation, code, anything!",
   "main": "./lib/generator.js",
   "bin": {
-    "asyncapi-generator": "./cli.js",
-    "ag": "./cli.js"
+    "asyncapi-generator": "cli.js",
+    "ag": "cli.js"
```
Comment (derberg) [inline package.json:8]: was this change necessary?

File: package.json [hunk 2]
```diff
@@ -118,5 +115,10 @@
       "@semantic-release/npm",
       "@semantic-release/github"
     ]
+  },
+  "directories": {
+    "doc": "docs",
+    "lib": "lib",
+    "test": "test"
```
Comment (derberg) [inline package.json:122]: do we need this? especially that `test` is in `.gitignore`?

File: test/test-project/test-global.test.js [hunk 3]
```diff
@@ -10,30 +10,31 @@ const strThatShouldBeMissing = '<div class="text-sm text-gray-700 mb-2">Correlat
 const templateName = '@asyncapi/html-template';
 const tempOutputResults = '../temp/integrationTestResult';
 const fileToCheck = 'index.html';
+const templateNotfoundInGlobalPackage = require('../../lib/logMessages');
```
Comment (derberg) [inline test/test-project/test-global.test.js:13]: don't you think it is better to call it the same as in the main code? I just have a feeling that later it introduces a kind of duplication in the naming `templateNotfoundInGlobalPackage.templateNotFound`

File: lib/logMessages.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
+const template_Install_Started_Msg ='Template installation started because you passed --install flag';
```
Comment (derberg) [inline lib/logMessages.js:1]: this const should not be created, it is as you can see used only in tests, instead, in tests you should use the log function that you created. Something like:
```js
installationDebugMessage(logMessages.TEMPLATE_INSTALL_FLAG_MSG);
```

File: lib/generator.js [hunk 5]
```diff
@@ -350,23 +352,23 @@ class Generator {
...
             path: pkgPath
           });
         } catch (e) {
-          log.debug(`Unable to resolve template location at ${pkgPath}. Package is not available locally`, e);
+          log.debug(logMessage.packageNotAvailable(pkgPath), e);
           // We did our best. Proceed with installation...
         }
       }
 
       const debugMessage = force ? 'because you passed --install flag' : 'because the template cannot be found on disk';
```
Comment (derberg) [inline lib/generator.js:368]: these 2 should also be moved to `logMessages`

**Review by derberg — CHANGES_REQUESTED** [review]
Looks good! thanks. I just had few comments/suggestions.

Some general comments:
- pay attention to consistency, this is very important in a project where more than one person participates. It is all about avoiding confusion
- for global constants, it is common to use all caps approach. Wherever you have constants like `node_modules_Install` or similar, please refactor to `NODE_MODULES_INSTALL`
- in few log messages dot `.` is missing

Additional info: 
In AsyncAPI we follow conventional commits specification to automate releases. We also do squash and merge at the end and this means:
- PR title prefixes are important, in your case you just do `refactor:`, `feat:` would trigger minor release of the package
- We do not care much about your commit messages for individual commits as we at the end squash them into one commit that is 1:1 same as the PR title
- To make review smoother just remember to never rebase with force, every change is a new commit so we can see what has changed in the PR since we had a look for the last time.


Please let me know if you need something from me

File: test/test-project/test-global.test.js (hunk 3 above)
Comment (vishesh13byte) [reply test/test-project/test-global.test.js:13]: Ya , I think I have complicated this a bit just gonna change it as `logMessages`.

File: package.json (hunk 1 above)
Comment (vishesh13byte) [reply package.json:8]: This change is not necessary ,changing it back to original.

File: lib/logMessages.js (hunk 4 above)
Comment (vishesh13byte) [reply lib/logMessages.js:1]: Umm okay , I will just simply return it from that method.

**Comment by vishesh13byte** [comment]
@derberg I have made all the requested changes 
> the installation message is not a constant anymore ,its is passed as a variable 
> `TEMPLATE_INSTALL_FLAG_MSG` and `TEMPLATE_INSTALL_DISK_MSG` have been moved to `logMessages`
>  I have corrected the missing dots `.` at the end of messages except the message returned by `installationDebugMessage` method as had I used a `.` there ,errors would have occurred in _generator.test.js_ file because it was expecting the message `expect(log.debug).toHaveBeenCalledWith('Template installation started because the template cannot be found on disk');`

Please review it again .
Thank you.

File: test/test-project/test-project.test.js [hunk 6]
```diff
@@ -53,8 +54,8 @@ describe('Testing if html was generated with proper version of the template', ()
 
     //we make sure that index.html file doesn't contain any infromation about correlationId because this feature was added in html-template 0.17.0 while this test uses template 0.16.0
     expect(isCorelationIdInHtml).toStrictEqual(true);
-    expect(console.log).toHaveBeenCalledWith(templateInstallStartedMsg);
-    expect(console.log).toHaveBeenCalledWith(`Version of used template is ${templateVersion}.`);
+    expect(console.log).toHaveBeenCalledWith(logMessage.installationDebugMessage(logMessage.TEMPLATE_INSTALL_FLAG_MSG));
+    expect(console.log).toHaveBeenCalledWith(logMessage.tempVersion(templateVersion));
```
Comment (derberg) [inline test/test-project/test-project.test.js:58]: `tempVersion` does not exist as far as I can see. strange that linter doesn't complain 🤔

File: test/test-project/test-project.test.js [hunk 7]
```diff
@@ -83,28 +84,28 @@ describe('Testing if html was generated with proper version of the template', ()
...
     file = await readFile(path.join(outputDir, fileToCheck), 'utf8');
     isCorelationIdInHtml = file.includes(strThatShouldBeMissing);
 
     //we make sure that index.html file doesn't contain any infromation about correlationId because this feature was added in html-template 0.17.0 while this test uses template 0.16.0
     expect(isCorelationIdInHtml).toStrictEqual(false);
-    expect(console.log).toHaveBeenCalledWith(`Template sources taken from ${localHtmlTemplate}.`);
-    expect(console.log).toHaveBeenCalledWith(`Version of the template is ${templateVersion}.`);
-    expect(console.log).toHaveBeenCalledWith('Remember that your local template must have its own node_modules installed first, \"npm install\" is not triggered by the generator.');
+    expect(console.log).toHaveBeenCalledWith(logMessage.templateSource(localHtmlTemplate));
+    expect(console.log).toHaveBeenCalledWith(logMessage.tempVersion(templateVersion));
```
Comment (derberg) [inline test/test-project/test-project.test.js:95]: same here about `tempVersion`

File: test/test-project/test-project.test.js [hunk 8]
```diff
@@ -83,28 +84,28 @@ describe('Testing if html was generated with proper version of the template', ()
...
     file = await readFile(path.join(outputDir, fileToCheck), 'utf8');
     isCorelationIdInHtml = file.includes(strThatShouldBeMissing);
 
     //we make sure that index.html file doesn't contain any infromation about correlationId because this feature was added in html-template 0.17.0 while this test uses template 0.16.0
     expect(isCorelationIdInHtml).toStrictEqual(false);
-    expect(console.log).toHaveBeenCalledWith(templateInstallStartedMsg);
-    expect(console.log).toHaveBeenCalledWith(`Version of used template is ${ templateVersion}.`);
-    expect(console.log).toHaveBeenCalledWith('Installation of template located on disk technically means symlink creation betweed node_modules of the generator and template sources. Your local template must have its own node_modules, "npm install" is not triggered.');
+    expect(console.log).toHaveBeenCalledWith(logMessage.installationDebugMessage(logMessage.TEMPLATE_INSTALL_FLAG_MSG));
+    expect(console.log).toHaveBeenCalledWith(logMessage.tempVersion(templateVersion));
```
Comment (derberg) [inline test/test-project/test-project.test.js:108]: same here about `tempVersion`

**Review by derberg — COMMENTED** [review]
I think you missed few `log.debug` usages, from utils and their tests for sure.

Please search entire project for `log.debug` usage

**Comment by vishesh13byte** [comment]
@derberg I have extracted all the log messages now .

Thank you.

File: test/test-project/test-global.test.js [hunk 9]
```diff
@@ -10,30 +10,31 @@ const strThatShouldBeMissing = '<div class="text-sm text-gray-700 mb-2">Correlat
...
+
     const file = await readFile(path.join(outputDir, fileToCheck), 'utf8');
     const isCorelationIdInHtml = file.includes(strThatShouldBeMissing);
 
     //we make sure that index.html file doesn't contain any information about correlationId because this feature was added in html-template 0.17.0 while this test uses template 0.16.0
     expect(isCorelationIdInHtml).toStrictEqual(false);
     //we make sure that logs indicate that global package was used
-    expect(console.log).toHaveBeenCalledWith(`${templateName} not found in local dependencies but found it installed as a global package.`);
+    expect(console.log).toHaveBeenCalledWith(logMessage.templateNotFound(templateName));
     expect(console.log).toHaveBeenCalledWith('Version of the template is 0.16.0.');
```
Comment (derberg) [inline test/test-project/test-global.test.js:38]: one log message left

File: test/test-project/test-project.test.js [hunk 10]
```diff
@@ -11,39 +11,40 @@ const strThatShouldBeMissing = '<div class="text-sm text-gray-700 mb-2">Correlat
...
     const file = await readFile(path.join(outputDir, fileToCheck), 'utf8');
     const isCorelationIdInHtml = file.includes(strThatShouldBeMissing);
 
     //we make sure that index.html file doesn't contain any infromation about correlationId because this feature was added in html-template 0.17.0 while this test uses template 0.16.0
     expect(isCorelationIdInHtml).toStrictEqual(false);
     //we make sure that logs do not indicate that new installation is started
-    expect(console.log).not.toHaveBeenCalledWith(templateInstallStartedMsg);
+    expect(console.log).not.toHaveBeenCalledWith(logMessage.installationDebugMessage(logMessage.TEMPLATE_INSTALL_FLAG_MSG));
     expect(console.log).toHaveBeenCalledWith(`Template sources taken from ${path.join(__dirname, 'node_modules', templateName)}.`);
     expect(console.log).toHaveBeenCalledWith('Version of the template is 0.16.0.');
```
Comment (derberg) [inline test/test-project/test-project.test.js:41]: another 2 log messages are still here

**Comment by derberg** [comment]
@vishesh13byte 3 logs messages left for refactor

**Comment by vishesh13byte** [comment]
> @vishesh13byte 3 logs messages left for refactor

I have done all those changes , I did not know that I needed to change that becuase it contained a constant `0.16.0`. Thanks for bearing with me ,the way you just guided and helped me out is amazing.

Thank you.

**Comment by derberg** [comment]
no worries, happy to help

you needed to change `Version of the template is 0.16.0.` because of 2 things:
- in the library you were using a helper that was already having a different message `Version of used template is ${ver}.`. This is why tests were failing on a CI
- the task was about replacing all those strings everywhere to use only helpers and constants 😄 

Thanks a lot for all the fixes, all checks are green now. I'll have a final look

File: lib/logMessages.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+function templateNotFound(templateName) {
+  return `${templateName} not found in local dependencies but found it installed as a global package.`;
+} 
+
+function packageNotAvailable(pkgPath) {
+  return `Unable to resolve template location at ${pkgPath}. Package is not available locally.`;
+}
+
+function installationDebugMessage(debugMessage) {
+  return `Template installation started ${debugMessage}`;
```
Comment (derberg) [inline lib/logMessages.js:26]: please add `.` at the end of the sentence, same in line 1 and 3. Let us be consistent and have dot in all the log messages. With this simple change, you will see why it was worth the effort to do the refactor. Next time we do any improvements in log messages texts, we just do them here in one file, no need to touch main library code and tests.

**Review by derberg — COMMENTED** [review]
@vishesh13byte one last tiny thing and I'm ready to approve and merge

File: lib/logMessages.js (hunk 11 above)
Comment (vishesh13byte) [reply lib/logMessages.js]: Yes ,you are absolutely right and this one change is significant as it maintains the consistency .There is so much I have learnt while working on thi issue and consistency in code is definitely up there.

Thank you.

**Review by derberg — APPROVED** [review]
Great work!
